File: app/src/main/java/com/raredev/vcspace/adapters/FileListAdapter.java
package com.raredev.vcspace.adapters;

import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.material.imageview.ShapeableImageView;
import com.google.android.material.textview.MaterialTextView;
import com.raredev.vcspace.R;
import com.raredev.vcspace.databinding.LayoutFileItemBinding;
import com.raredev.vcspace.models.FileExtension;
import com.raredev.vcspace.ui.viewmodel.FileListViewModel;
import java.io.File;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class FileListAdapter extends RecyclerView.Adapter<FileListAdapter.VH> {

  private FileListViewModel viewModel;
  private FileListener fileListener;

  public FileListAdapter(FileListViewModel viewModel) {
    this.viewModel = viewModel;
  }

  @NonNull
  @Override
  public VH onCreateViewHolder(ViewGroup parent, int arg1) {
    return new VH(
        LayoutFileItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false));
  }

  @Override
  public void onBindViewHolder(VH holder, int position) {
    holder.itemView.setAnimation(
        AnimationUtils.loadAnimation(holder.itemView.getContext(), android.R.anim.fade_in));

    File file = viewModel.getFiles().get(position);

    if (!viewModel.getSelectedFiles().contains(file)) {
      holder.itemView.setBackgroundColor(Color.TRANSPARENT);
    } else {
      holder.itemView.setBackgroundColor(Color.parseColor("#87B7E5FF"));
    }

    holder.tv_name.setText(file.getName());
    if (file.isFile()) {
      holder.img_icon.setImageResource(FileExtension.Factory.forFile(file).getIcon());
    } else {
      holder.img_icon.setImageResource(R.drawable.ic_folder);
    }

    if (position == 0) {
      holder.img_icon.setImageResource(R.drawable.ic_folder);
    }

    holder.itemView.setOnClickListener(
        (v) -> {
          if (viewModel.getSelectedFiles().contains(file)) {
            holder.itemView.setBackgroundColor(Color.TRANSPARENT);
            viewModel.removeSelectFile(file);
            return;
          }
          if (fileListener != null) {
            fileListener.onFileClick(file, v);
          }
        });

    holder.itemView.setOnLongClickListener(
        (v) -> {
          if (fileListener != null) {
            fileListener.onFileLongClick(file, v);
          }
          return true;
        });

    holder.img_icon.setOnClickListener(
        v -> {
          if (position == 0) {
            return;
          }
          if (viewModel.getSelectedFiles().contains(file)) {
            holder.itemView.setBackgroundColor(Color.TRANSPARENT);
            viewModel.removeSelectFile(file);
          } else {
            holder.itemView.setBackgroundColor(Color.parseColor("#87B7E5FF"));
            viewModel.selectFile(file);
          }
        });
  }

  @Override
  public int getItemCount() {
    return viewModel.getFiles().size();
  }

  public void refreshFiles() {
    viewModel.getSelectedFiles().clear();
    notifyDataSetChanged();
  }

  public boolean isFilesSelected() {
    return !viewModel.getSelectedFiles().isEmpty();
  }

  public List<File> getSelectedFiles() {
    return viewModel.getSelectedFiles();
  }

  public void setFileListener(FileListener listener) {
    this.fileListener = listener;
  }

  public interface FileListener {
    void onFileClick(File file, View v);

    void onFileLongClick(File file, View v);
  }

  public class VH extends RecyclerView.ViewHolder {
    ShapeableImageView img_icon;
    MaterialTextView tv_name;

    public VH(LayoutFileItemBinding binding) {
      super(binding.getRoot());
      img_icon = binding.imgIcon;
      tv_name = binding.fileName;
    }
  }
}

File: app/src/main/java/com/raredev/vcspace/adapters/FileListAdapter.kt
package com.raredev.vcspace.adapters

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.raredev.vcspace.databinding.LayoutFileItemBinding
import com.raredev.vcspace.providers.FileIconProvider
import com.raredev.vcspace.resources.R
import java.io.File
import java.text.SimpleDateFormat

class FileListAdapter(private val listener: OnFileClickListener) :
  RecyclerView.Adapter<FileListAdapter.VH>() {

  private var files: List<File> = emptyList()

  inner class VH(internal val binding: LayoutFileItemBinding) :
    RecyclerView.ViewHolder(binding.root)

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
    return VH(LayoutFileItemBinding.inflate(LayoutInflater.from(parent.context), parent, false))
  }

  @SuppressLint("SimpleDateFormat")
  override fun onBindViewHolder(holder: VH, position: Int) {
    holder.binding.apply {
      val file = files[position]

      val iconRes: Int =
        if (file.isFile) {
          FileIconProvider.findFileIconResource(file)
        } else R.drawable.ic_folder

      icon.setImageResource(iconRes)
      name.text = file.name
      info.text =
        root.context.getString(
          R.string.file_modified_in,
          SimpleDateFormat("yy/MM/dd").format(file.lastModified())
        )

      root.setOnClickListener { listener.onFileClickListener(file) }
      root.setOnLongClickListener { listener.onFileLongClickListener(file, it) }
    }
  }

  override fun getItemCount(): Int {
    return files.size
  }

  fun submitList(newFiles: List<File>) {
    val diffResult = DiffUtil.calculateDiff(FileDiffCallback(this.files, newFiles))
    this.files = newFiles
    diffResult.dispatchUpdatesTo(this)
  }

  interface OnFileClickListener {
    fun onFileClickListener(file: File)

    fun onFileLongClickListener(file: File, view: View): Boolean
  }

  inner class FileDiffCallback(private val oldList: List<File>, private val newList: List<File>) :
    DiffUtil.Callback() {
    override fun getOldListSize(): Int {
      return oldList.size
    }

    override fun getNewListSize(): Int {
      return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
      return oldList[oldItemPosition].name == newList[newItemPosition].name
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
      return oldList[oldItemPosition].path == newList[newItemPosition].path
    }
  }
}
